// @ngx-cache/platform-browser/src/local-storage-cache.service.js
import { isPlatformBrowser } from '@angular/common';
import { Inject, PLATFORM_ID } from '@angular/core';
var LocalStorageCacheService = (function () {
    function LocalStorageCacheService(platformId) {
        this.platformId = platformId;
        if (!isPlatformBrowser(platformId))
            throw new Error('LocalStorageCacheService is not supported outside `browser` platform!');
    }
    Object.defineProperty(LocalStorageCacheService.prototype, "isEnabled", {
        get: function () {
            if (!isPlatformBrowser(this.platformId))
                return false;
            try {
                localStorage.setItem('test', 'test');
                localStorage.removeItem('test');
                return true;
            }
            catch (e) {
                return false;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LocalStorageCacheService.prototype, "keys", {
        get: function () {
            if (!this.isEnabled)
                return undefined;
            var res = [];
            Object.keys(localStorage).forEach(function (key) {
                res.push(key);
            });
            return res;
        },
        enumerable: true,
        configurable: true
    });
    LocalStorageCacheService.prototype.setItem = function (key, value) {
        if (!this.isEnabled)
            return false;
        try {
            localStorage.setItem(key, JSON.stringify(value));
            return true;
        }
        catch (e) {
            return false;
        }
    };
    LocalStorageCacheService.prototype.getItem = function (key) {
        if (!this.isEnabled)
            return undefined;
        var value = localStorage.getItem(key);
        return value ? JSON.parse(value) : undefined;
    };
    LocalStorageCacheService.prototype.removeItem = function (key, wild) {
        if (wild === void 0) { wild = false; }
        if (!this.isEnabled)
            return;
        localStorage.removeItem(key);
        if (wild)
            for (var _i = 0, _a = this.keys; _i < _a.length; _i++) {
                var item = _a[_i];
                if (item.indexOf(key) === 0)
                    localStorage.removeItem(item);
            }
    };
    LocalStorageCacheService.prototype.clear = function () {
        if (!this.isEnabled)
            return;
        localStorage.clear();
    };
    return LocalStorageCacheService;
}());
export { LocalStorageCacheService };
LocalStorageCacheService.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] },] },
]; };
